sdk/core/Azure.Core/src/Response.cs

```diff
@@ -154,5 +138,150 @@ internal static void DisposeStreamIfNotBuffered(ref Stream? stream)
+        /// <returns></returns>
+        /// <exception cref="NotImplementedException"></exception>
+        public override BinaryData ReadContent(CancellationToken cancellationToken = default)
+        {
+            // Derived types should provide an implementation that allows caching
+            // to improve performance.
+            if (ContentStream is null)
+            {
+                return s_EmptyBinaryData;
+            }
+
+            if (ContentStream is MemoryStream memoryStream)
+            {
+                return memoryStream.TryGetBuffer(out ArraySegment<byte> segment) ?
+                    new BinaryData(segment.AsMemory()) :
+                    new BinaryData(memoryStream.ToArray());
+            }
+
+            BufferedContentStream bufferStream = new();
+
+            Stream? contentStream = ContentStream;
+            contentStream.CopyTo(bufferStream, cancellationToken);
+            contentStream.Dispose();
+
+            bufferStream.Position = 0;
+            ContentStream = bufferStream;
+
+            BinaryData content = BinaryData.FromStream(bufferStream);
+            bufferStream.Position = 0;
+
+            return content;
+        }
+
+        /// <summary>
+        /// TBD.
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="NotImplementedException"></exception>
+        public override async ValueTask<BinaryData> ReadContentAsync(CancellationToken cancellationToken = default)
+        {
+            // Derived types should provide an implementation that allows caching
+            // to improve performance.
+            if (ContentStream is null)
+            {
+                return s_EmptyBinaryData;
+            }
+
+            if (ContentStream is MemoryStream memoryStream)
+            {
+                return memoryStream.TryGetBuffer(out ArraySegment<byte> segment) ?
+                    new BinaryData(segment.AsMemory()) :
+                    new BinaryData(memoryStream.ToArray());
+            }
+
+            BufferedContentStream bufferStream = new();
+
+            Stream? contentStream = ContentStream;
+            await contentStream.CopyToAsync(bufferStream, cancellationToken).ConfigureAwait(false);
+            contentStream.Dispose();
+
+            bufferStream.Position = 0;
+            ContentStream = bufferStream;
+
+            BinaryData content = BinaryData.FromStream(bufferStream);
+            bufferStream.Position = 0;
+
+            return content;
+        }
+
+        private class BufferedContentStream : MemoryStream { }
+
+        #region Private implementation subtypes of abstract Response types
+        private class AzureCoreResponse<T> : Response<T>
+        {
+            public AzureCoreResponse(T value, Response response)
+                : base(value, response) { }
+        }
+
+        internal class AzureCoreDefaultResponse : Response
```

Comment:
What's the purpose of this class if it adds no functionality to `Response` and we lose the true abstract behavior?

Reply:
It allows us to validate that the `PipelineResponse` passed to the base type constructor is non-null.  If you can suggest another approach that allows that, I am happy to take it -- I think I had a TODO comment to see if I could find another way on this type until recently, so I agree it's not the optimal approach.

Reply:
Added comments to explain in https://github.com/Azure/azure-sdk-for-net/pull/42234